ql/src/java/org/apache/hadoop/hive/ql/cache/results/QueryResultsCache.java

```diff
@@ -783,13 +755,11 @@ private boolean hasSpaceForCacheEntry(CacheEntry entry, long size) {
 
   private CacheEntry findEntryToRemove() {
     // Entries should be in LRU order in the keyset iterator.
-    Set<CacheEntry> entries = lru.keySet();
     synchronized (lru) {
-      for (CacheEntry removalCandidate : entries) {
-        if (removalCandidate.getStatus() != CacheEntryStatus.VALID) {
-          continue;
+      for (CacheEntry removalCandidate : lru) {
+        if (removalCandidate.getStatus() == CacheEntryStatus.VALID) {
+          return removalCandidate;
         }
-        return removalCandidate;
```

Comment:
Grab the `keySet` _inside_ of the synchronized block to avoid ConcurrentModification. Remove use of 'continue' keyword.